server/src/test/java/org/apache/calcite/avatica/server/RemoteUserExtractHttpServerTest.java

```diff
@@ -0,0 +1,175 @@
+ */
+package org.apache.calcite.avatica.server;
+
+import org.apache.calcite.avatica.ConnectionSpec;
+import org.apache.calcite.avatica.jdbc.JdbcMeta;
+import org.apache.calcite.avatica.remote.AuthenticationType;
+import org.apache.calcite.avatica.remote.Driver;
+import org.apache.calcite.avatica.remote.LocalService;
+
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Properties;
+import java.util.concurrent.Callable;
+
+import javax.servlet.http.HttpServletRequest;
+
+import static org.hamcrest.core.StringContains.containsString;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.fail;
+
+/**
+ * Test class for HTTP Basic authentication.
+ */
+public class RemoteUserExtractHttpServerTest extends HttpAuthBase {
+  private static final Logger LOG = LoggerFactory.getLogger(RemoteUserExtractHttpServerTest.class);
+
+  private static final ConnectionSpec CONNECTION_SPEC = ConnectionSpec.HSQLDB;
+  private static HttpServer server;
+  private static String url1;
+  private static String url2;
+
+  @BeforeClass public static void startServer() throws Exception {
+
+    final String userPropertiesFile = BasicAuthHttpServerTest.class
+            .getResource("/auth-users.properties").getFile();
+    assertNotNull("Could not find properties file for basic auth users", userPropertiesFile);
+
+    // Create a LocalService around HSQLDB
+    final JdbcMeta jdbcMeta = new JdbcMeta(CONNECTION_SPEC.url,
+            CONNECTION_SPEC.username, CONNECTION_SPEC.password);
+    LocalService service = new LocalService(jdbcMeta);
+
+    HandlerFactory factory = new HandlerFactory();
+    AvaticaHandler avaticaHandler = factory.getHandler(service, Driver.Serialization.PROTOBUF, null,
+        new AvaticaServerConfiguration() {
+          @Override public AuthenticationType getAuthenticationType() {
+            return AuthenticationType.BASIC;
+          }
+
+          @Override public String getKerberosRealm() {
+            return null;
+          }
+
+          @Override public String getKerberosPrincipal() {
+            return null;
+          }
+
+          @Override public boolean supportsImpersonation() {
+            return true;
+          }
+
+          @Override public <T> T doAsRemoteUser(String remoteUserName, String remoteAddress,
+                                                Callable<T> action) throws Exception {
+
+            if (remoteUserName.equals("USER4")) {
+              throw new RuntimeException("USER4 is a disallowed user");
+            } else if (remoteUserName.equals("USER2")) {
+              return action.call();
+            } else {
+              throw new RuntimeException("Unknown user.");
+            }
+          }
+
+          @Override public RemoteUserExtractor getRemoteUserExtractor() {
+            return new RemoteUserExtractor() {
```

Comment:
1. I tried to mock the real usage by adding ?useDoAs=true param in the url indicating if the server expects doAs or not, is that something you are expecting?
2. One weird thing is if I use request.getRemoteUser() it will return null in this case, so I just "hardcode" to return USER2 in the extractor. Not sure if using basic authentication influences this because this extraction behavior only happens in spnego scenario, but the test cases are basic auth and I manually add avaticaServerConfiguration to force execute this code.